src/librustc_ast_passes/ast_validation.rs

```diff
@@ -826,28 +826,36 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                 if let Unsafe::Yes(span) = unsafety {
                     struct_span_err!(
                         self.session,
-                        item.span,
+                        vec![span, self_ty.span],
                         E0197,
                         "inherent impls cannot be unsafe"
                     )
                     .span_label(span, "unsafe because of this")
+                    .span_label(self_ty.span, "inherent impl for this type")
                     .emit();
                 }
-                if polarity == ImplPolarity::Negative {
-                    self.err_handler().span_err(item.span, "inherent impls cannot be negative");
+                if let ImplPolarity::Negative(span) = polarity {
+                    self.err_handler().span_err(span, "inherent impls cannot be negative");
```

Comment:
Would prefer to have `struct_span_err(...)` here with the `sm.def_span(item.span)`, the "inherent impl for this type", and then a `span_label` for the `!`.